File: python/src/ambertrace/providers/anthropic/interceptor.py
"""Anthropic SDK interception via monkey-patching."""

import functools
import logging
import time
import uuid
from typing import Any, Callable, Optional

from ambertrace.providers.anthropic.collector import AnthropicCollector
from ambertrace.providers.base import BaseInterceptor
from ambertrace.transport import get_transport

logger = logging.getLogger(__name__)


class AnthropicInterceptor(BaseInterceptor):
    """Handles monkey-patching of Anthropic SDK methods.

    This class wraps Anthropic client methods to intercept calls and collect traces.
    The wrapping is transparent - all original behavior is preserved exactly.
    """

    def __init__(self) -> None:
        """Initialize interceptor."""
        self._original_sync_create: Optional[Callable] = None
        self._original_async_create: Optional[Callable] = None
        self._is_patched = False
        self._collector = AnthropicCollector()

    def get_provider_name(self) -> str:
        """Return provider name."""
        return "anthropic"

    def patch(self) -> None:
        """Apply monkey-patches to Anthropic SDK methods.

        Wraps:
        - anthropic.Anthropic.messages.create (sync)
        - anthropic.AsyncAnthropic.messages.create (async)
        """
        if self._is_patched:
            logger.debug("Anthropic already patched, skipping")
            return

        try:
            import anthropic

            # Patch sync method
            if hasattr(anthropic, "Anthropic"):
                try:
                    # Navigate to messages.create
                    # We need to patch at the resources level
                    if hasattr(anthropic.resources, "Messages"):
                        messages_class = anthropic.resources.Messages

                        if hasattr(messages_class, "create"):
                            self._original_sync_create = messages_class.create
                            messages_class.create = self._wrap_sync_create(
                                self._original_sync_create
                            )
                            logger.debug("Patched Anthropic.messages.create (sync)")
                        else:
                            logger.warning("Could not find create method on Messages class")
                    else:
                        logger.warning("Could not find Messages resources")

                except Exception as e:
                    logger.error(f"Failed to patch sync Anthropic method: {e}", exc_info=True)

            # Patch async method
            if hasattr(anthropic, "AsyncAnthropic"):
                try:
                    # Navigate to async messages.create
                    if hasattr(anthropic.resources, "AsyncMessages"):
                        async_messages_class = anthropic.resources.AsyncMessages

                        if hasattr(async_messages_class, "create"):
                            self._original_async_create = async_messages_class.create
                            async_messages_class.create = self._wrap_async_create(
                                self._original_async_create
                            )
                            logger.debug("Patched AsyncAnthropic.messages.create (async)")
                        else:
                            logger.warning("Could not find create method on AsyncMessages class")
                    else:
                        logger.warning("Could not find AsyncMessages resources")

                except Exception as e:
                    logger.error(f"Failed to patch async Anthropic method: {e}", exc_info=True)

            self._is_patched = True
            logger.info("Anthropic SDK patching completed")

        except ImportError:
            logger.warning("Anthropic SDK not found - skipping Anthropic tracing")
        except Exception as e:
            logger.error(f"Unexpected error during Anthropic patching: {e}", exc_info=True)

    def unpatch(self) -> None:
        """Remove monkey-patches and restore original Anthropic methods."""
        if not self._is_patched:
            logger.debug("Anthropic not patched, skipping unpatch")
            return

        try:
            import anthropic

            # Restore sync method
            if self._original_sync_create is not None:
                try:
                    if hasattr(anthropic.resources, "Messages"):
                        anthropic.resources.Messages.create = self._original_sync_create
                        logger.debug("Restored original Anthropic sync create method")
                except Exception as e:
                    logger.error(f"Failed to restore Anthropic sync method: {e}", exc_info=True)

            # Restore async method
            if self._original_async_create is not None:
                try:
                    if hasattr(anthropic.resources, "AsyncMessages"):
                        anthropic.resources.AsyncMessages.create = self._original_async_create
                        logger.debug("Restored original Anthropic async create method")
                except Exception as e:
                    logger.error(f"Failed to restore Anthropic async method: {e}", exc_info=True)

            self._is_patched = False
            self._original_sync_create = None
            self._original_async_create = None
            logger.info("Anthropic SDK unpatching completed")

        except Exception as e:
            logger.error(f"Unexpected error during Anthropic unpatching: {e}", exc_info=True)

    def is_patched(self) -> bool:
        """Check if patches are currently applied."""
        return self._is_patched

    def _wrap_sync_create(self, original_method: Callable) -> Callable:
        """Create wrapper for sync messages.create method.

        Args:
            original_method: Original Anthropic method to wrap

        Returns:
            Wrapped method that collects traces
        """

        @functools.wraps(original_method)
        def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            """Wrapper that intercepts sync Anthropic calls."""
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            try:
                # Call original method
                response = original_method(self_instance, *args, **kwargs)

                # Collect trace (non-blocking, never raises)
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=response,
                        error=None,
                    )

                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as e:
                    # Never let trace collection errors affect user code
                    logger.error(f"Error collecting Anthropic trace {trace_id}: {e}", exc_info=True)

                # Return original response unchanged
                return response

            except Exception as e:
                # Collect trace with error information
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=None,
                        error=e,
                    )

                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            transport.send_trace(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Anthropic error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )

                # Re-raise original exception to preserve user experience
                raise

        return wrapper

    def _wrap_async_create(self, original_method: Callable) -> Callable:
        """Create wrapper for async messages.create method.

        Args:
            original_method: Original async Anthropic method to wrap

        Returns:
            Wrapped async method that collects traces
        """

        @functools.wraps(original_method)
        async def wrapper(self_instance: Any, *args: Any, **kwargs: Any) -> Any:
            """Wrapper that intercepts async Anthropic calls."""
            trace_id = str(uuid.uuid4())
            start_time = time.time()

            try:
                # Await original method
                response = await original_method(self_instance, *args, **kwargs)

                # Collect trace (non-blocking, never raises)
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=response,
                        error=None,
                    )

                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            # Use async send for async context
                            transport.send_trace_async(trace_dict)
                except Exception as e:
                    # Never let trace collection errors affect user code
                    logger.error(f"Error collecting Anthropic trace {trace_id}: {e}", exc_info=True)

                # Return original response unchanged
                return response

            except Exception as e:
                # Collect trace with error information
                try:
                    trace_dict = self._collector.collect_trace(
                        trace_id=trace_id,
                        start_time=start_time,
                        request_kwargs=kwargs,
                        response=None,
                        error=e,
                    )

                    if trace_dict:
                        transport = get_transport()
                        if transport:
                            # Use async send for async context
                            transport.send_trace_async(trace_dict)
                except Exception as trace_error:
                    logger.error(
                        f"Error collecting Anthropic error trace {trace_id}: {trace_error}",
                        exc_info=True,
                    )

                # Re-raise original exception to preserve user experience
                raise

        return wrapper

File: typescript/src/providers/anthropic/interceptor.ts
/**
 * Anthropic SDK interceptor using prototype manipulation.
 *
 * Intercepts messages.create() to collect traces transparently.
 */

import { randomUUID } from 'crypto';
import type { BaseInterceptor } from '../base';
import { AnthropicCollector } from './collector';
import { getTransport } from '../../transport';
import { getConfig } from '../../config';

/**
 * Intercepts Anthropic SDK message creation calls.
 */
export class AnthropicInterceptor implements BaseInterceptor {
  private collector = new AnthropicCollector();
  private isCurrentlyPatched = false;
  private originalCreate: ((...args: unknown[]) => unknown) | null = null;

  getProviderName(): string {
    return 'anthropic';
  }

  patch(): void {
    if (this.isCurrentlyPatched) {
      return; // Already patched
    }

    try {
      // Try to import Anthropic SDK
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const anthropic = require('@anthropic-ai/sdk');

      // Access the Messages class
      const MessagesClass = anthropic.Anthropic?.Messages;
      if (!MessagesClass || !MessagesClass.prototype) {
        throw new Error('Anthropic SDK structure not recognized');
      }

      const prototype = MessagesClass.prototype;

      // Store original create method
      this.originalCreate = prototype.create;

      if (typeof this.originalCreate !== 'function') {
        throw new Error('messages.create is not a function');
      }

      // Create wrapper
      const originalMethod = this.originalCreate;
      const self = this;

      prototype.create = function wrappedCreate(...args: unknown[]) {
        const config = getConfig();

        // If tracing is disabled, call original method
        if (!config?.enabled) {
          return originalMethod.apply(this, args);
        }

        // Extract request arguments
        // Anthropic SDK: create(body, options)
        const requestArgs = (args[0] as Record<string, unknown>) ?? {};

        // Generate trace ID and start time
        const traceId = randomUUID();
        const startTime = Date.now();

        try {
          // Call original method
          const result = originalMethod.apply(this, args);

          // Check if result is a promise (async)
          if (result && typeof (result as Promise<unknown>).then === 'function') {
            return (result as Promise<unknown>)
              .then((response) => {
                // Collect trace on success
                self.collectAndSendTrace(traceId, startTime, requestArgs, response);
                return response;
              })
              .catch((error: Error) => {
                // Collect trace on error, then re-throw
                self.collectAndSendTrace(traceId, startTime, requestArgs, undefined, error);
                throw error;
              });
          }

          // Synchronous result (unlikely for Anthropic SDK)
          self.collectAndSendTrace(traceId, startTime, requestArgs, result);
          return result;
        } catch (error) {
          // Collect trace on exception, then re-throw
          self.collectAndSendTrace(
            traceId,
            startTime,
            requestArgs,
            undefined,
            error as Error
          );
          throw error;
        }
      };

      this.isCurrentlyPatched = true;

      if (getConfig()?.debug) {
        console.log('[AmberTrace] Anthropic SDK patched successfully');
      }
    } catch (error) {
      console.error('[AmberTrace] Failed to patch Anthropic SDK:', error);
      throw error;
    }
  }

  unpatch(): void {
    if (!this.isCurrentlyPatched || !this.originalCreate) {
      return;
    }

    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const anthropic = require('@anthropic-ai/sdk');
      const MessagesClass = anthropic.Anthropic?.Messages;

      if (MessagesClass?.prototype) {
        MessagesClass.prototype.create = this.originalCreate;
      }

      this.isCurrentlyPatched = false;
      this.originalCreate = null;

      if (getConfig()?.debug) {
        console.log('[AmberTrace] Anthropic SDK unpatched');
      }
    } catch (error) {
      console.error('[AmberTrace] Failed to unpatch Anthropic SDK:', error);
    }
  }

  isPatched(): boolean {
    return this.isCurrentlyPatched;
  }

  private collectAndSendTrace(
    traceId: string,
    startTime: number,
    requestArgs: Record<string, unknown>,
    response?: unknown,
    error?: Error
  ): void {
    try {
      // Collect trace
      const trace = this.collector.collectTrace(traceId, startTime, requestArgs, response, error);

      if (trace) {
        // Send trace asynchronously
        const transport = getTransport();
        transport.sendTrace(trace);
      }
    } catch (err) {
      // Never throw from trace collection
      if (getConfig()?.debug) {
        console.error('[AmberTrace] Error collecting/sending trace:', err);
      }
    }
  }
}
